### PR TITLE
Set metadata for vtable-related loads

### LINKED ISSUE #39992: LLVM pessimises vtable loads.
This code contains two offset calculations and two function pointer loads:

```rust
#![crate_type="lib"]

pub struct Wrapper<T: ?Sized> {
    x: usize,
    y: usize,
    t: T
}

pub trait Foo {
    fn foo(&self);
}

pub fn test(foo: &Wrapper<Foo>) {
    foo.t.foo();
    foo.t.foo();
}
```

[Playpen](https://play.rust-lang.org/?gist=0153602ac2b5d6060f56c08aa7ee831e)

LLVM should be able to remove all but the first occurrences via common subexpression elimination, but for some reason it reloads from the vtable pointer each time. The only reasonable explanation I can think of is that LLVM assumes that the vtable pointer may alias some pointer that the called function manipulates.

### PR DESCRIPTION
**Pull request by Aatch**
Give LLVM much more information about vtable pointers. Without the extra
information, LLVM has to be rather pessimistic about vtables, preventing
a number of obvious optimisations.

* Makes the vtable pointer argument noalias and readonly.
* Marks loads of the vtable pointer as nonnull.
* Marks load from the vtable with `!invariant.load` metadata.

Fixes #39992

### CODE REVIEW

**Comment by Aatch** [comment]
LLVM IR before and after this patch: https://gist.github.com/Aatch/fe555f251de910d7e765a11cbb5d78d4

**Comment by arielb1** [comment]
unchanged test:
```
C:\projects\rust\src/test\codegen\function-arguments.rs:124:11: error: expected string not found in input
// CHECK: @trait_borrow(i8* nonnull, void (i8*)** nonnull)
          ^
C:\projects\rust\build\x86_64-pc-windows-gnu\test\codegen\function-arguments.ll:266:62: note: scanning from here
define internal void @str(i8* noalias nonnull readonly, i64) unnamed_addr #1 {
                                                             ^
C:\projects\rust\build\x86_64-pc-windows-gnu\test\codegen\function-arguments.ll:276:22: note: possible intended match here
define internal void @trait_borrow(i8* nonnull, void (i8*)** noalias nonnull readonly) unnamed_addr #1 {
```

**Comment by gnzlbg** [comment]
@Aatch do you have a benchmark/examples for this somewhere? Like to be able to see the "before" and "after" on the generated IR? LLVM devirtualization has improved a lot in the last year (still pretty basic though), but it would be nice to have a couple of "benchmarks" for cases in which we expect a backend to be able to devirtualize calls. That might also help filling LLVM bugs for the cases in which this is not the case, or discover situations in which we are not passing meta-data correctly.

File: src/librustc_trans/base.rs [hunk 1]
```diff
@@ -472,8 +472,15 @@ pub fn load_fat_ptr<'a, 'tcx>(
...
-    // FIXME: emit metadata on `meta`.
-    let meta = b.load(get_meta(b, src), alignment.to_align());
+    let meta = get_meta(b, src);
+    let meta_ty = val_ty(meta);
+    // If the 'meta' field is a pointer, it's a vtable, so use load_nonnull
+    // instead
+    let meta = if meta_ty.element_type().kind() == llvm::TypeKind::Pointer {
+        b.load_nonnull(meta, None)
+    } else {
+        b.load(meta, None)
```
Comment (bjorn3) [inline src/librustc_trans/base.rs]: Missing alignment?

File: src/librustc_trans/abi.rs [hunk 2]
```diff
@@ -506,7 +506,11 @@ impl FnType {
                 if let Some(inner) = rust_ptr_attrs(ty, &mut data) {
                     data.attrs.set(ArgAttribute::NonNull);
                     if ccx.tcx().struct_tail(inner).is_trait() {
+                        // vtables can be safely marked non-null, readonly
+                        // and noalias.
                         info.attrs.set(ArgAttribute::NonNull);
+                        info.attrs.set(ArgAttribute::ReadOnly);
```
Comment (nikic) [inline src/librustc_trans/abi.rs]: Isn't `readonly` a [function attribute](http://llvm.org/docs/LangRef.html#function-attributes)? What does this mean when applied to arguments?

File: src/librustc_trans/base.rs (hunk 1 above)
Comment (arielb1) [inline src/librustc_trans/base.rs]: None = ABI alignment. Vtables are ABI aligned.

File: src/librustc_trans/abi.rs (hunk 2 above)
Comment (Aatch) [inline src/librustc_trans/abi.rs]: When applied to a pointer argument, it means that the function will only read from it. That allows LLVM to assume that the value behind the pointer won't change across a call to that function.

**Comment by dpc** [comment]
Please take a look at [my last comment in #39992](https://github.com/rust-lang/rust/issues/39992#issuecomment-284129795) about tight loops and this.

**Pull request merged**